Source/WebCore/dom/Element.cpp

```diff
@@ -2735,232 +2735,313 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
 }
 
-static bool isValidDirValue(const AtomString& value)
+static TextDirectionState elementTextDirectionState(const Element& element)
 {
-    return parseTextDirection(value) != TextDirectionDirective::Invalid;
+    return parseTextDirectionState(element.attributeWithoutSynchronization(dirAttr));
 }
 
-static bool elementAffectsDirectionality(const Element& element)
+static bool elementHasValidTextDirectionState(const Element& element)
 {
-    return is<HTMLBDIElement>(element) || isValidDirValue(element.attributeWithoutSynchronization(dirAttr));
+    return elementTextDirectionState(element) != TextDirectionState::Undefined;
 }
 
-static bool elementAffectsDirectionality(const Node& node)
+static bool elementHasAutoTextDirectionState(const Element& element)
 {
-    auto* element = dynamicDowncast<Element>(node);
-    return element && elementAffectsDirectionality(*element);
+    auto textDirectionState = elementTextDirectionState(element);
+    return textDirectionState == TextDirectionState::Auto || (is<HTMLBDIElement>(element) && textDirectionState == TextDirectionState::Undefined);
 }
 
-static void setHasDirAutoFlagRecursively(Node* firstNode, bool flag, Node* lastNode = nullptr)
+static void updateHasDirAutoFlagForSubtree(Node& firstNode, bool flag)
 {
-    firstNode->setSelfOrPrecedingNodesAffectDirAuto(flag);
+    firstNode.setSelfOrPrecedingNodesAffectDirAuto(flag);
 
-    RefPtr node = firstNode->firstChild();
+    for (RefPtr node = firstNode.firstChild(); node; ) {
+        auto* element = dynamicDowncast<Element>(*node);
 
-    while (node) {
-        if (elementAffectsDirectionality(*node)) {
-            if (node == lastNode)
-                return;
-            node = NodeTraversal::nextSkippingChildren(*node, firstNode);
+        if (element && (is<HTMLBDIElement>(*element) || elementHasValidTextDirectionState(*element))) {
+            node = NodeTraversal::nextSkippingChildren(*node, &firstNode);
             continue;
         }
+
         node->setSelfOrPrecedingNodesAffectDirAuto(flag);
-        if (node == lastNode)
-            return;
-        node = NodeTraversal::next(*node, firstNode);
+        node = NodeTraversal::next(*node, &firstNode);
     }
 }
 
-bool Element::hasDirectionAuto() const
+static void updateElementHasDirAutoFlag(Element& element, TextDirectionState textDirectionState)
 {
-    const AtomString& direction = attributeWithoutSynchronization(dirAttr);
-    return (hasTagName(bdiTag) && !isValidDirValue(direction)) || equalLettersIgnoringASCIICase(direction, "auto"_s);
+    RefPtr parent = element.parentOrShadowHostElement();
+    element.protectedDocument()->setIsDirAttributeDirty();
+
+    switch (textDirectionState) {
+    case TextDirectionState::LTR:
+    case TextDirectionState::RTL:
+        if (element.selfOrPrecedingNodesAffectDirAuto() || (parent && parent->selfOrPrecedingNodesAffectDirAuto()))
+            updateHasDirAutoFlagForSubtree(element, false);
```

Comment:
Can we define an enum class instead of using boolean here? It's unclear what "false" and "true" means in this context.

Reply:
Instead of creating a new enum, I am passing `textDirectionState` instead and set the selfOrPrecedingNodesAffectDirAuto flag to true only when the `textDirectionState != TextDirectionState::Auto`